control-base/src/main/java/fi/nls/oskari/control/layer/DataProviderHandler.java

```diff
@@ -17,19 +25,97 @@
+	private DataProviderService dataProviderService = ServiceFactory.getDataProviderService();
+	private OskariLayerService mapLayerService = ServiceFactory.getMapLayerService();
 
 	@Override
 	public void handleGet(ActionParameters params) throws ActionException {
-		
+
 		final int id = params.getRequiredParamInt(PARAM_ID);
-		
+
 		try {
-			DataProvider d = service.find(id);
+			DataProvider d = dataProviderService.find(id);
 			ResponseHelper.writeResponse(params, d.getAsJSON());
 		} catch (Exception e) {
 			throw new ActionException("Data provider query failed", e);
 		}
 	}
+
+	@Override
+	public void handleDelete(ActionParameters params) throws ActionException {
+
+		final int id = params.getRequiredParamInt(PARAM_ID);
+		final boolean deleteLayers = params.getRequiredParamBoolean(PARAM_DELETE_LAYERS);
+
+		if (!dataProviderService.hasPermissionToUpdate(params.getUser(), id)) {
+			throw new ActionDeniedException("Unauthorized user tried to remove data provider id=" + id);
+		}
+
+		if (deleteLayers) {
+			deleteDataProviderAndLayers(id, params);
+		} else {
+			deleteDataProvider(id, params);
+		}
+	}
+
+	/**
+	 * Method deletes data provider. Data provider is removed from map layers
+	 * belonging to the data provider prior delete.
+	 *
+	 * @param int              id
+	 * @param ActionParameters params
+	 * @throws ActionException
+	 */
+	private void deleteDataProvider(int id, ActionParameters params) throws ActionException {
+		try {
+			DataProvider dataProvider = dataProviderService.find(id);
+			if (dataProvider == null) {
+				throw new ActionParamsException("Dataprovider not found with id " + id);
+			}
+			List<OskariLayer> layers = mapLayerService.findByDataProviderId(id);
+			layers.stream().forEach(layer -> {
+				layer.removeDataprovider(id);
+				mapLayerService.update(layer);
+			});
+			dataProviderService.delete(id);
+			AuditLog.user(params.getClientIp(), params.getUser()).withParam("id", dataProvider.getId())
+					.withParam("name", dataProvider.getName(PropertyUtil.getDefaultLanguage()))
+					.deleted(AuditLog.ResourceType.DATAPROVIDER);
+
+			// write deleted organization as response
+			ResponseHelper.writeResponse(params, dataProvider.getAsJSON());
+		} catch (Exception e) {
+			throw new ActionException("Couldn't delete data provider with id:" + id, e);
+		}
+	}
+
+	/**
+	 * Method deletes data provider and map layers which belong to the data
+	 * provider.
+	 *
+	 * @param int              id
+	 * @param ActionParameters params
+	 * @throws ActionException
+	 */
+	private void deleteDataProviderAndLayers(int id, ActionParameters params) throws ActionException {
+		try {
+			DataProvider organization = dataProviderService.find(id);
+			if (organization == null) {
+				throw new ActionParamsException("Dataprovider not found with id " + id);
+			}
+			// cascade in db will handle that layers are deleted
+			dataProviderService.delete(id);
```

Comment:
Wonder if we should write audit log for layers being deleted this way. I'm not sure but maybe the auditlog call below could indicate that layers were removed as well. You can add .withMsg() next to the .withParam() call. Maybe with the handleDelete refactoring we could AuditLog on handleDelete to make it clear we are logging the deletion.

Reply:
Added deleted layer names to audit logging.